**Table of Content** 
- [Blood DHS](#blood-dhs)
- [H3K27ac](#h3k27ac)
- [Phylo](#phylo)
- [TOPMed](#topmed)
- [ClinVar](#clinvar)
- [OMIM](#omim)
- [gnomAD](#gnomad)
- [GWAS](#gwas)
- [Fine-mapping](#fine-mapping)
- [MSigDB](#msigdb)
- [DHS\_index](#dhs_index)
- [SCREEN](#screen)
- [chromHMM](#chromhmm)
- [motif](#motif)
- [Classic papers used database above](#classic-papers-used-database-above)

#### Blood DHS 
 - Current project no collection of Blood DNase-seq. In order to integrate blood celltype into our study with RNA-seq and HiTrAC, public DNase-seq data of blood cells are collected;
 - plan to download from ENCODE: See "encode_blood_dhs.md" for details; 
 - celltypes with highquality library included (all from PMBC): 
	1. B cell (no stimulation, no memory)
	2. CD4+ T cell (no stimulation)
	3. CD8+ T cell (no stimulation)
	4. Monocyte (CD14+CD18-)
	5. Natural killer (NK)
 -  The downloaded BAM file contains duplicated reads (per the metadata of BAM, duplicates retained), I performed additional step to remove duplicated-reads and also check the QC to met my cutting in DNase-seq pipeline (Note: To save space, after filtering, original BAM deleted). 
    - see `QC.py`;
    - only PE library used; 
    - low SPOT score, poor cluster in PCA result; 
 - peaks are called using MACS3 and filtered by fold-change 2.5 ("fc2.5"); 
 - briefing about PMBC: 
   - hematopoietic stem cell (HSC, CD34 marker) contains myeloid stem cell (MSC) and lymphoid stem cell (LSC);
   - myeloid stem cell can differentiate blood red cells, platelets, neutrophil, and monocyte (terminal to macrophage), important for innate immunity; 
   - lymphoid stem cell can differentiate into NK cell with large granular (for innate immunity), and T or B with small lymphocyte (for adaptive immunity); 
   - CD4 and CD8 are T cells, CD4 proliferate to T helper cells (Th1, Th2, Th17), and T helper can release cytokines to CD8, which can proliferate to cytotoxic T cells;
   - B cells terminate to plasma cell, generating huge amount of antibodies or memory B cells, and it requires T cell's help to be fully activated; 

#### H3K27ac 
- For active enhancers, it is marked by H3K27ac; 
- Get from [here](https://chip-atlas.org/search); 
- access metadata ("experimentList.tab"); https://github.com/inutano/chip-atlas/wiki#tables-summarizing-metadata-and-files 
- Since DNase-seq and Hi-TrAC data both indicated that some celltypes are more similar to each other (e.g., different tissues' endothelial cells are similar), I can then collect: 
   - all endothelial cell's H3k27ac data, merge them together as endothelial enhancer pool; 
   - all smooth muscle cells' H3K27ac data, merge them together as SMC enhancer pool; 
   - 

#### Phylo 
- See [download](https://hgdownload.soe.ucsc.edu/goldenpath/hg38/)
- phyloP and phastCons are conservation scores for human genome by performing multiple sequennce alignment to **other vetebrates**.
- PhastCons
	- hg38.phastCons470way.bw ("conservation score" indicates as the probability that each nucleotide belongs to a conserved element).  
	Note: it only detects conserved elements/regions; 
			longer conserved region rather than per-nucleotide variation;
			in a range [0-1] (closer to 1 = more conserved)

- PhyloP100
  - 100 vertebrate species; 
	- hg38.phyloP100way.bw ("conservation score" is transformed using -log(p), where P indicates the probability of the conserveness of a nucleotide site) 
	Note: it can detect both conserved and accelerated evolution;
	- contains both positive and negative values (> 0, strong evolutionary constraints, more conserved, more functional importance).
  - in a range from -20 to +10, higher postive values (+6 to +10) reflect high conservation, lower negative values (-5 to -10) reflect strong acceleration. 
  - about "way":
  	- 100way: 99 vetebrate align against human genome;
  	- 470way: 469 vetebrate align against human genome. 



#### TOPMed
- see home page [here](https://topmed.nhlbi.nih.gov/about);
- see [publication](https://doi.org/10.1038/s41586-021-03205-y);
  - 53,831 diverse genomes sequenced; 
  - 410 million SNP, INDELs detected (note that at each position, different alternative alleles may be identified); 
  - 97% of vairants identified have frequencies <1%; 46% are sinfletons (present in one individual only); 
  - 
- compare to 1000 Genome Project, HapMap3, and dbSNP, I decided to use TOPMed for human genetic variation resources, for the following reasons: 
  - Deep coverage (~30-40x) for high-quality variant call; 
  - Utilizing diverse populations, and uncovered rare variants; 
  - Generated by NHLBI with a focus on heart, lung, blood and sleep disorders (match to what I intended to do); 
- may be useful for the calculation of **nucleotide diversity** (a measure of genetic variation within a population, different from Phylo, which is among species); 
  -  about nucleotide diversity (π): 
  - high  π -> high genetic diversity (high mutation rates and recombination);
  - low  π -> low genetic diversity (in bottlenecked or inbred populations); 
  - used to compare the evoluationary pressure across genes; 
- Genotype data of [Freeze 8](https://legacy.bravo.sph.umich.edu/freeze8/hg38/downloads) can be accessed via BRAVO;  
  - VCF file for each individual chromosome. and to combine all chromosome VCF files using bcftools:
  - `bcftools merge chrom/*vcf.gz -o BRAVO_TOPMed_Freeze_8.vcf.gz -O z --threads 12 -W --force-samples`
  - -W automatically index the output file 
  - --force-samples: resolved duplicate sample names 



#### ClinVar 
- [ClinVar](ftp.ncbi.nih.gov/pub/clinvar/vcf_GRCh38/) contains variants related to phenotype of disease of clinical sigificance or benign; 
- For gene-disease relationship, see "gene_condition_source_id.txt"; 
- 


#### OMIM 
- Online Mendelian inheritance in Man; 
- The collected gene-disease relationships are also archieved by ClinVar. 
- OMIM is better than ClinVar for following reasons: 
  - Documented in greater detail (e.g., molecular basis known labeled as (3));
  - It includes more genes and its potential phenotypic effects; 
  - The phenotype (disease) is concatenated when associated with the same gene; 
  - More importantly, the inheritance mode (e.g., autosomal recessive) for the disease is clarified; 
- OMIM gene-disease prefix meaning: 
  - 'Asterisk': 17530 (gene, w/wo phenotype relationship), 
  - 'Number Sign': 6970 (descriptive entry, e.g., phenotype wo gene relationship), 
  - nan: 1732 (descriptive entry, e.g., phenotype of its mendelian basis not clearly established), 
  - 'Percent': 1499 (descriptive entry, e.g., confirmed phenotype with unclear molecular basis), 
  - 'Caret': 1371 (no longer exists), 
  - 'Plus': 12 (descriptive of gene and phenotype)


#### gnomAD 
- genome Aggregation Database: aggregate human exome and genome sequencing data; 
- Downloaded from UCSC track on its download server; 
  - [pliByGene.bb](https://genome.ucsc.edu/cgi-bin/hgTables)
  - [missenseByGene.bb](https://hgdownload.soe.ucsc.edu/gbdb/hg38/gnomAD/pLI/)
- LOEUF (Loss of function constraint)
- pLI is the probability of belonging in the haploinsufficient class; 
- Haploinsufficient genes: pLI >= 0.9; 
- Tolerant genes: pLI <= 0.1; 


#### GWAS 
- [GWAS Catalog database](https://www.ebi.ac.uk/gwas/docs/file-downloads) is for human genome-wide association studies; 
- version used: v1.0.2 (contains variant and trait ontology annotation); 
- to compare the dataset quality, I downloaded the variants used [here](https://www.sciencedirect.com/science/article/pii/S0092867413012270?via%3Dihub), what I found is: 
	-  GWAS catalog contains most of the variants used in that study, and the trait of specific variant in association is also in agreement. 
- about SNP prune: 
  - causal SNPs related to the specific trait is unknown, and it can be any variations with linkage disequilibruim of the SNP; 
- versions: 
  - v1.0 is the standard GWAS catalog info;
  - v1.0.2 with added ontology annotations, GWAS Catalog study accession numbers and genotype tech (preferred with ontology annotation);
  - v1.0.2.1 with ontology annotations, GWAS Catalog study accession numbers and genotype tech, cohort IDs, and full summary statistics;
  - use v1.0.2 as reference (optimal information); 
- filtering: 
  - I cleaned up v1.0.2 raw tsv; 
  - filter replicated variant for the same trait (see "clean"); 
  - when the same variant identified multiple times (from multiple studies), the variant with the most significant p-value then retained (see "clean.remove_replicates"); 
  - see "gwas.py"; 
- Test on enrichment analysis wo/ prune SNPs: 
  - For SIL of nCAEC, SNPs of phosphatidylinositol (PI) enriched; 
  - checking on the SNPs for PI measurement, I found that cluster of SNPs overlapping with SIL contribute to the enrichment (e.g., chr11:62 Mb, chr19: 54 Mb); 
- about LDSC: 
  - format explained: https://github.com/bulik/ldsc/wiki/LD-File-Formats
  - 

#### Fine-mapping 
- See [here](https://www.finucanelab.org/data); 
- 


#### MSigDB 
- GO annotation for [human/mouse](https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp);



#### DHS_index 
- DHS index from [publication](https://doi.org/10.1038/s41586-020-2559-3) (733 biosamples and 438 cell and tissue types, described over 3.6 million DHS);
  - Downloaded from [here](https://zenodo.org/records/3838751); 
    - see "DHS_Index_and_Vocabulary_hg38_WM20190703.txt.gz";
  - about DHS index and its vocabulary: 
	- method used [NMF](https://github.com/Altius/Vocabulary) (non-negative matrix factorization);
	- a total of 3,591,899 DHS included, classified into 16 components; 
	- NMF used to decompose components for all DHS (the biological context of each component was identified by the most strongly associated biosample, and the description of TF recognition sites);

#### SCREEN 
- [see](https://screen.encodeproject.org/);
- current version: cCREs v3; 
- New version 4: https://screen.wenglab.org/ 
- For all cCRE: GRCh38-cCREs_v4.bed; 
  - promoter like signatures (PLS): 200 bp of TSS and high DNase and H3K4me3 signals
  - enhancer like signatures (ELS): pELS: proximal ELS (< 2 kb of TSS), distal ELS (> 2 kb of TSS)
  - CTCF: high DNase and CTCF and low H3K4me3 and H3K27ac
  - other: GRCh38-cCREs.DNase-H3K4me3.bed, with high DNase-H3K4me3 fall out of 200 bp of TSS (low H3K27ac); 

#### chromHMM 
- public available data for the assignment of chromtain state for specific celltypes; 
- For 25 chromatin states, [see](https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/imputed12marks/jointModel/final/) 
- For 15 chromatin states, [see](https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/) 
- celltypes collected: 
  - HUVEC: E122 
  - LF: E128
  - EM: E059 
  - EK: E057 
- 5 chromatin marks assayed in all epigenomes (H3K4me3, H3K4me1, H3K36me3, H3K27me3, H3K9me3);
- 15 core states 
	STATE NO.	MNEMONIC	DESCRIPTION	COLOR NAME	COLOR CODE
	1	TssA	Active TSS	Red	255,0,0
	2	TssAFlnk	Flanking Active TSS	Orange Red	255,69,0
	3	TxFlnk	Transcr. at gene 5' and 3'	LimeGreen	50,205,50
	4	Tx	Strong transcription	Green	0,128,0
	5	TxWk	Weak transcription	DarkGreen	0,100,0
	6	EnhG	Genic enhancers	GreenYellow	194,225,5
	7	Enh	Enhancers	Yellow	255,255,0
	8	ZNF/Rpts	ZNF genes & repeats	Medium Aquamarine	102,205,170
	9	Het	Heterochromatin	PaleTurquoise	138,145,208
	10	TssBiv	Bivalent/Poised TSS	IndianRed	205,92,92
	11	BivFlnk	Flanking Bivalent TSS/Enh	DarkSalmon	233,150,122
	12	EnhBiv	Bivalent Enhancer	DarkKhaki	189,183,107
	13	ReprPC	Repressed PolyComb	Silver	128,128,128
	14	ReprPCWk	Weak Repressed PolyComb	Gainsboro	192,192,192
	15	Quies	Quiescent/Low	White	255,255,255

#### motif 
- TF and Co-factors Download from [AnimalTFDB4](https://guolab.wchscu.cn/AnimalTFDB4/#/Download)
  - see "Hs_TF_Cof.txt"; 
- TF footprint predicted from DHS, see [paper](https://doi.org/10.1038/s41586-020-2528-x);
  - download from https://zenodo.org/records/3905306;
  - Use deeply sequenced DNase-seq for 243 cell-types and to delineate about 4.5 million genomic elements that for TF occupancy; 
  - ~80% assigned to >= TF, 20% assigned to single TF; 46% to single TF with lower ranked alternatives (by TF motif searching); 
  - see `collapsed_motifs_overlapping_consensus_footprints_hg38.bed`; 
  - check it later???
- ENCODE motif based de novo searching (based on TF-specific ChIP-seq data), see [paper](https://academic.oup.com/nar/article/42/5/2976/1050297); 
  - Downloaded using [encodemotif R package](https://rdrr.io/bioc/motifbreakR/man/encodemotif.html), see "encodemotif.txt" (motif for specific TF, total of 703 TF with total 2065 motifs);
  - To locate motif coordinates in the genome, used FIMO: 
	`fimo --skip-matched-sequence --text encodemotif.meme $hg38`
	Note: there are 28 motifs have no coordinates identified. The resulting coordinates file transferred into bed format (**2037** bed files).  
- For processed motifs coordinates for each factor: 
  - download it [here](https://compbio.mit.edu/encode-motifs/);
  - see "matches.txt.gz" for motif matches on hg19 (renamed to "matches_hg19.bed4"); 
  - as the genome version is hg19, change to hg38; 
	- `liftOver matches_hg19.bed /data/jim4/Reference/human/liftOver/hg19ToHg38.over.chain.gz matches_hg38.bed unmapped`
  - there are 602 TF factors in file; 

#### Classic papers used database above 
- Landscape of cohesin-mediated chromatin loops in the human genome (https://doi.org/10.1038/s41586-020-2151-x). 
    ENCODE TF motif db (hg19)
    ClinVar
    GWAS catalog 
    haploinsufficient genes 

- Index and biological spectrum of human DNase I hypersensitive sites (https://doi.org/10.1038/s41586-020-2559-3).
    Genome repeat classes from UCSC 
    TOPMed from Bravo (from SNV to nucleotide diversity scores)
    PhyloP (100way)
    MSigDB collection (for canonical pathways)
    KEGG 
	GWAS from UK Biobank by Neale lab 
	HapMap3 

- Occupancy maps of 208 chromatinassociated proteins in one human cell type (https://doi.org/10.1038/s41586-020-2023-4)
	CIS_BP & JASPAR (motif similarity analysis)
	Human Protein Atlas (RNA expression for liver specific gene)
	IDEAS chromatin segmention 
	Genomic evolutionary rate profiling (GERP, from Sidow lab at Stanford)
	PIQ for DNase footprinting 
	SOMatic package for SOM analysis 

- 